### PR TITLE
Define API

### PR DESCRIPTION
**Pull request by rnburn**
This PR proposes an API for Lua. It builds off of @james-callahan's Kong API https://github.com/Kong/opentracing-lua, but removes the implementation portions that are usually put into basictracer projects (e.g. https://github.com/opentracing/basictracer-python).

The goal with the Lua API is to support these use cases:
* Working with tracers completely written in Lua (for example, Kong's Zipkin tracer https://github.com/sfhardman/kong-plugin-zipkin).
* Loading existing C++ tracers as plugins and using from Lua.
* For C++ applications that embed Lua, bridging and interoperating with the C++ tracers the application may already use.

Given the lack of direct interface support in Lua, the API isn't required as a dependency as it is with the other languages; but conforming tracers are required to implement the methods as documented.

I also wrote a [bridge-tracer](https://github.com/rnburn/lua-bridge-tracer) that implements the API and allows you to use existing C++ tracers through the dynamic loading interface.

Cc @james-callahan, @ror6ax, @isaachier for review.

### CODE REVIEW

**Comment by isaachier** [comment]
Wow really awesome work @rnburn! I am glad to see someone did this work for Lua. Maybe this effort will prove worthwhile in other languages as well.

File: opentracing/span.lua
Comment (james-callahan) [inline opentracing/span.lua]: add semicolon?

File: opentracing/span_context.lua
Comment (james-callahan) [inline opentracing/span_context.lua]: Mixed tabs and spaces.

**Review by james-callahan — CHANGES_REQUESTED** [review]
Thanks for putting in the work!

Yet I'm not sure what the use of all this is (due to lua not needing "interfaces"):
  - is it a template project for others starting opentracing implementations?
  - should it act as a "base class" somehow?

**Comment by rnburn** [comment]
@james-callahan - it's only serving as documentation and as a specification for what methods conforming OT Lua tracers need to implement.

I don't think it needs to provide anything that's used directly; though at some point it might make sense to add in a mocktracer and these [constants](https://github.com/opentracing/opentracing-python/blob/master/opentracing/ext/tags.py) that are usually defined.

File: opentracing/span.lua
Comment (rnburn) [reply opentracing/span.lua]: Added.

File: opentracing/span_context.lua
Comment (rnburn) [reply opentracing/span_context.lua]: Fixed.

**Comment by james-callahan** [comment]
Looks good now; if you want to squash down your commits to remove intermediary fixes I can merge.

**Review by james-callahan — APPROVED** [review]

**Comment by rnburn** [comment]
@james-callahan - I set up the repo to squash on merge (the default for the other opentracing languages).

Let's keep it open a little longer. I'm going to post it on the OT gitter channels to see if anyone has more feedback.

**Comment by james-callahan** [comment]
> @james-callahan - I set up the repo to squash on merge (the default for the other opentracing languages).

:( I really hate squash-on-merging. it breaks commit signing, in addition to making it hard to pull apart large pull requests. At least it doesn't appear to be the default.

> Let's keep it open a little longer. I'm going to post it on the OT gitter channels to see if anyone has more feedback.

I feel like for starting a project it's better to have *something* and have people file PRs to fix it rather than for feedback on the initial implementation.
I'm happy for it to be merged as-is.

**Comment by tedsuo** [comment]
Hi @james-callahan and @rnburn, I'm going to merge this since commentary has died down.

To explain the squash-and-merge philosophy: for API repos, we want master to have a clean commit history, where every commit on master to represents a usable version. Otherwise, there's a risk that a user could land on something broken, or ideas that we tried and then backed away from. But we don't want to require all intermediary commits on a PR branch to have this requirement. Hence squash-and-merge. As long as the squashed commits are linked to the original PRs, the git history remains for someone who wants to investigate the details. Not perfect but that's the reasoning. We should become even stricter about our process as the APIs mature and become stable.

For early, unstable versions like this, I think we should move fast to merge changes into master, we shouldn't build up a giant pile of changes on a single PR.

**Comment by tedsuo** [comment]
@rnburn btw I noticed there was no version number yet. Can we do SEMVER and start this as `v0.1`? I'll merge once that's added.

**Comment by rnburn** [comment]
I set `_VERSION = '0.1.0'`

**Comment by isaachier** [comment]
Isn't there a canonical way to set `_VERSION`? Does luarocks provide the version from the rockspec to users?

Edit: Clarify

**Comment by isaachier** [comment]
Actually never mind. Looks like some packages do supply it in both the rockspec and the lua source. See [here](https://github.com/craigbarnes/lua-gumbo/blob/master/gumbo.lua#L94) for an example.

**Comment by tedsuo** [comment]
LGTM. Seems like there's a smear of versioning approaches in Lua. @isaachier we can always refactor how we are exposing/versioning.

**Comment by daurnimator** [comment]
> we want master to have a clean commit history, where every commit on master to represents a usable version.

I entirely agree. But I think it should always be up to the pull-request author to make that clean history, never the reviewer/merger.

> But we don't want to require all intermediary commits on a PR branch to have this requirement.

There should never be intermediary commits in a PR; the author should amend them as they improve the PR.

**Comment by james-callahan** [comment]
> we want master to have a clean commit history, where every commit on master to represents a usable version.

I entirely agree. But I think it should always be up to the pull-request author to make that clean history, never the reviewer/merger.

> But we don't want to require all intermediary commits on a PR branch to have this requirement.

There should never be intermediary commits in a PR; the author should amend them as they improve the PR.

**Comment by yurishkuro** [comment]
>  The author should amend them as they improve the PR.

This only creates unnecessary friction, for both maintainers and contributors. I almost never encounter a situation where it's imperative that commits in the PR must be preserved in master - in most cases it's just a sign that the PR is doing too many things and should be split.

**Pull request merged**
